Clustering/src/KMeans.java

```diff
@@ -1,10 +1,117 @@
 /**
  * K-Means clustering
  */
+import java.lang.Math;
+import java.util.ArrayList;
+
 public class KMeans  extends Clustering{
 
+    /**
+     * Clusters data via the K-means Algorithm
+     */
     @Override
     public int[] cluster(double[][] data, int numClusters){
+        int cycles =0 ;
+        double[][] centroids = initCentroids(data, numClusters);
+        while(!hasConverged(cycles)){
+            ArrayList<Integer> labels = getLabels(centroids, data);
+            centroids = getCentroids(labels, data, numClusters);
+            cycles++;
+        }
         return null;
     }
+
+    /**
+     * returns true if K-means has converged
+     */
+    private boolean hasConverged(int cycles) {
+        return (cycles >= 200);
+    }
+
+    /**
+     * initializes centroids to random datapoints
+     */
+    public double[][] initCentroids(double[][] data, int numClusters){
+        double[][] centroids = new double[numClusters][];
+
+        for (int centIter = 0; centIter < numClusters; centIter++) {
+            //pick a random datapoint from the sample and make it a centroid
+            centroids[centIter] = data[(int)(Math.random() * data.length)];
+        }
+        return centroids;
+    }
+
+    /**
+     * assigns the closest centroid to every datapoint
+     */
+    private ArrayList<Integer> getLabels(double[][] centroids, double[][] data) {
+        ArrayList<Integer> labels = new ArrayList<>();
+
+        for (double[] point: data) {
+            labels.add(labelPoint(point, centroids));
+        }
+        return labels;
+    }
+
+    /**
+     * finds centroid closest to given point and labels that point
+     */
+    private Integer labelPoint(double[] point, double[][] centroids) {
```

Comment:
I would use `int` instead so we avoid any hard-to-debug reference errors

Reply:
done